framework/src/org/checkerframework/common/jaifinference/JaifInferenceUtils.java

```diff
@@ -0,0 +1,485 @@
+     */
+    private static List<AnnotationMirror> annotationsToIgnore =
+            new ArrayList<AnnotationMirror>();
+
+    /**
+     * @param classSymbol is the symbol representing the class containing the method.
+     * @param methodElt is the method element.
+     * @param atf is the annotated type factory.
+     * @return the ATM of methodElt in the .jaif file of the class with symbol
+     * classSymbol, or null if it can't be found.
+     */
+    public static AnnotatedTypeMirror getMethodReturnTypeInJaif(
+            ClassSymbol classSymbol, ExecutableElement methodElt,
+            AnnotatedTypeFactory atf) {
+        if (classSymbol == null) return null; // Static block.
+        String jaifPath = JAIF_FILES_PATH + classSymbol.flatname.toString().
+                replaceAll("\\.", "/") + ".jaif";
+        try {
+            AClass clazz = getJaifClass(classSymbol, jaifPath, new AScene());
+            if (clazz == null) return null;
+            String methodName = JVMNames.getJVMMethodName(methodElt);
+            AMethod method = clazz.methods.get(methodName);
+            if (method != null && method.returnType != null) {
+                Set<Annotation> prevAnnos = method.returnType.tlAnnotationsHere;
+                return setOfAnnotationsToATM(prevAnnos, atf, false,
+                            methodElt.getReturnType());
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    /**
+     * @param classSymbol is the symbol representing the class containing the method.
+     * @param fieldNode is the node representing the field.
+     * @param atf is the annotated type factory.
+     * @return the ATM of fieldNode in the .jaif file of the class with symbol
+     * classSymbol, or null if it can't be found.
+     */
+    public static AnnotatedTypeMirror getFieldTypeInJaif(
+            ClassSymbol classSymbol, FieldAccessNode fieldNode,
+            AnnotatedTypeFactory atf) {
+        if (classSymbol == null) return null; // Static block.
+        String jaifPath = JAIF_FILES_PATH + classSymbol.flatname.toString().
+                replaceAll("\\.", "/") + ".jaif";
+
+        AScene scene = new AScene();
+        try {
+            AClass clazz = getJaifClass(classSymbol, jaifPath, scene);
+            if (clazz == null) return null;
+            String fieldName = TreeUtils.getFieldName(fieldNode.getTree());
+            AField field = clazz.fields.get(fieldName);
+            if (field == null) {
+                return null;
+            }
+            Set<Annotation> annos = field.tlAnnotationsHere;
+            if (annos != null && annos.size() > 0) {
+                return setOfAnnotationsToATM(annos, atf, false, fieldNode.getType());
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    /**
+     * Updates the type of the field lhs in the .jaif file of the class with
```

Comment:
Does this routine rewrite the entire .jaif file every time any value in it changes?
It seems that it would be more efficient to update the scene throughout type-checking, and then to write each file only once at the end of type-checking.  What are the pros and cons of the alternate approach?